lib/functions/installs.bash

```diff
@@ -117,30 +117,199 @@ install_local_tool_versions() {
-    for plugin_name in $plugins_installed; do
-      local plugin_version_and_path
-      plugin_version_and_path="$(find_versions "$plugin_name" "$search_path")"
-
-      if [ -n "$plugin_version_and_path" ]; then
-        local plugin_version
-        some_tools_installed='yes'
-        plugin_versions=$(cut -d '|' -f 1 <<<"$plugin_version_and_path")
-        for plugin_version in $plugin_versions; do
-          install_tool_version "$plugin_name" "$plugin_version"
-        done
-      fi
-    done
+    plugins_installed=$(printf "%s" "$plugins_installed" | tr "\n" " " | awk '{$1=$1};1')
+    display_debug "install_local_tool_versions: plugins_installed='$plugins_installed'"
+    tools_installed=$(install_directory_tools_recursive "$search_path" "$plugins_installed")
   fi
 
-  if [ -z "$some_tools_installed" ]; then
+  if [ -z "$tools_installed" ]; then
     printf "Either specify a tool & version in the command\n"
     printf "OR add .tool-versions file in this directory\n"
     printf "or in a parent directory\n"
     exit 1
   fi
 }
 
+install_directory_tools_recursive() {
+  local search_path=$1
+  local plugins_installed=$2
+  local tools_installed=""
+
+  display_debug "install_directory_tools_recursive '$search_path': entered with plugins_installed='$plugins_installed'"
+
+  while [ "$search_path" != "/" ]; do
+    # install tools from files in current directory
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    tools_installed=$(install_directory_tools "$search_path" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$search_path': install_directory_tools returned tools_installed='$tools_installed'"
+
+    # terminate if all tools are installed
+    if [[ -n $(stringlist_a_subset_of_b "$plugins_installed" "$tools_installed") ]]; then
+      display_debug "install_directory_tools_recursive '$search_path': exiting because all known tools ($plugins_installed) are installed ($tools_installed)"
+      printf "%s\n" "$tools_installed"
+      return 0
+    fi
+
+    # go up a directory
+    search_path=$(dirname "$search_path")
+  done
+
+  # still got some tools to install...
+  # lets see if $ASDF_DEFAULT_TOOL_VERSIONS_FILENAME as actually
+  # an absolute path
+  if [ -f "$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME" ]; then
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    display_debug "attempting to treat \$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME as an absolute path: $ASDF_DEFAULT_TOOL_VERSIONS_FILENAME"
+    tools_installed=$(_install_directory_tools "" "$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME': install_directory_tools returned tools_installed='$tools_installed'"
+  fi
+
```

Comment:
there is no test for this, however this is behaviour that was in asdf already. I know because people at my work were using it like this.

Reply:
The docs are clear this is only the **filename** of the file - https://asdf-vm.com/manage/configuration.html#asdf-default-tool-versions-filename

It is not an absolute path to a file and shouldn't be treated like it is here (lines 180-181 should be all that's needed). Was there any existing code that treated it like this?

Reply:
happy to remove this.

Reply:
when I tested asdf on of our company repos I noticed that we were setting `ASDF_DEFAULT_TOOL_VERSIONS_FILENAME` to an absolute path and it was working with asdf. Then I tested my new change and it obviously stopped working. It seems like it could have been one of those "accident by design" things.